adding more heat terms #1254

Alex2448 opened the pull request:
## Summary of the discussion

This PR closes [Issue 1254](https://github.com/OpenEnergyPlatform/ontology/issues/1254).

## Type of change (CHANGELOG.md)

### Added
- `heat generating unit`
- <s>`plate heat exchanger`</s> transferred to #1432
- <s>`rotary heat exchanger`</s> transferred to #1432
- `heat plant`
- `heat transfer unit`
- <s>`boiler`</s> transferred to #1432
- `combustion-based heater`
- <s>`electrical heater`</s> transferred to #1432
- `geothermal heat unit`
- `geothermal heat plant`
- <s>`tube collector`</s> transferred to #1432
- <s>`flat-plate collector`</s> transferred to #1432
- `solar heat plant`
- `solar heat unit`

### Updated
- `power plant` --> more precise axioms
- `combined heat and power plant` --> change as a subclass of energy converting component
- `combined heat and power generating unit` --> change as a subclass of energy converting component
- <s>`solar thermal collector` --> improve definition, remove @en label</s> transferred to #1432

### Removed

## Workflow checklist

### Automation
* Closes #1254 

### PR-Assignee
- [x] 🐙 Follow the [Pull Request Workflow](https://github.com/OpenEnergyPlatform/ontology/wiki/Pull-request-workflow)
- [x] 📝 Update the [CHANGELOG.md](https://github.com/OpenEnergyPlatform/ontology/blob/dev/CHANGELOG.md)
- [x] 📙 Add #'s to `term tracker item`

### Reviewer
- [ ] 🐙 Follow the [Reviewer Guide](https://github.com/OpenEnergyPlatform/ontology/wiki/Pull-request-workflow#reviewer-guide-check-changes-introduced-by-a-pull-request)
- [ ] 🐙 Provided feedback and show sufficient appreciation for the work done

l-emele commented:
Somehow this PR did not get any review for weeks despite eight appointed reviewers.

I found the following problems: 
* `combined heat and power plant` is not anymore an `power plant`, not even in inferred hierarchy. But in my view, that contradicts the view of the domain. Maybe `power plant` needs to be converted to an equivalent class?
* Every `combined heat and power generating unit` should be also at least an inferred subclass of `power generating unit`. Maybe we should make `'power generating unit' EquivalentTo: `'energy transformation unit'` and ('has part' some 'generator')`?  Likewise, `heat generating unit` could be implemented as `'heat generating unit' EquivalentTo: 'energy transformation unit' and ('has part' some 'heat exchanger' or heater)`. Then `combined heat and power generating unit` would be an inferred subclass of both `power generating unit` and `heat generating unit`.

Alex2448 commented:
> Somehow this PR did not get any review for weeks despite eight appointed reviewers.
> 
> I found the following problems:
> 
> * `combined heat and power plant` is not anymore an `power plant`, not even in inferred hierarchy. But in my view, that contradicts the view of the domain. Maybe `power plant` needs to be converted to an equivalent class?
> * Every `combined heat and power generating unit` should be also at least an inferred subclass of `power generating unit`. Maybe we should make `'power generating unit' EquivalentTo: `'energy transformation unit'` and ('has part' some 'generator')`?  Likewise, `heat generating unit` could be implemented as `'heat generating unit' EquivalentTo: 'energy transformation unit' and ('has part' some 'heat exchanger' or heater)`. Then `combined heat and power generating unit` would be an inferred subclass of both `power generating unit` and `heat generating unit`.

I think the last part is a nice solution, does anyone have any objections? @OpenEnergyPlatform/oeo-domain-expert-energy-modelling 
Otherwise, I would implement it like l-emele suggested.

l-emele commented on src/ontology/edits/oeo-physical.omn:
```diff
@@ -9662,8 +9683,213 @@ pull request: https://github.com/OpenEnergyPlatform/ontology/pull/1363",
+
+    Annotations:
+        OEO_00040001 "http://openenergy-platform.org/ontology/oeo/oeo-physical"^^xsd:string,
+        <http://purl.obolibrary.org/obo/IAO_0000115> "A heat generating unit is an energy transformation unit that can contain a heat exchanger or a heater.",
```
Shorter and clearer:
```suggestion
        <http://purl.obolibrary.org/obo/IAO_0000115> "A heat generating unit is an energy transformation unit that contains a heat exchanger or a heater.",
```

l-emele commented on src/ontology/edits/oeo-physical.omn:
```diff
@@ -9662,8 +9683,213 @@ pull request: https://github.com/OpenEnergyPlatform/ontology/pull/1363",
+    SubClassOf:
+        OEO_00000210,
+        (OEO_00000503 some OEO_00000331)
+         and (OEO_00000531 value OEO_00000256)
```
This axiom should rather be: `uses some fluid`

Alex2448 commented:
> Somehow this PR did not get any review for weeks despite eight appointed reviewers.
> 
> I found the following problems:
> 
> * `combined heat and power plant` is not anymore an `power plant`, not even in inferred hierarchy. But in my view, that contradicts the view of the domain. Maybe `power plant` needs to be converted to an equivalent class?
> * Every `combined heat and power generating unit` should be also at least an inferred subclass of `power generating unit`. Maybe we should make `'power generating unit' EquivalentTo: `'energy transformation unit'` and ('has part' some 'generator')`?  Likewise, `heat generating unit` could be implemented as `'heat generating unit' EquivalentTo: 'energy transformation unit' and ('has part' some 'heat exchanger' or heater)`. Then `combined heat and power generating unit` would be an inferred subclass of both `power generating unit` and `heat generating unit`.

I implemented this now and also added made power plant and heater plant an equivalent class so that `combined heat and power plant` is inferred `power plant` and `heater plant`. So I added:
- for power plant: 
'energy transformation unit'
 and ('has part' some 'electricity grid component') and  ('has part' some 'power generating unit')
- for heat plant: 
'energy transformation unit'
 and ('has part' some 'heating grid component') and  ('has part' some 'heat generating unit')

l-emele reviewed (changes requested):
This implementation does not work.

In the dev branch, both `power generating unit` and `power plant` are also in inferred view direct subclasses of `energy transformation unit`:
![grafik](https://user-images.githubusercontent.com/36884905/204346645-03849a57-cee7-438f-9ab7-1ef903cadd14.png)

In your branch feature-1254-adding-more-heat-terms however, in inferred view, `power plant` is now a subclass of `power generating unit`:
![grafik](https://user-images.githubusercontent.com/36884905/204347532-86efc4ed-5f77-48aa-9ced-37d6e991bdb4.png)

But this cannot be true, as a `power plant` consists of one **or multiple** power generating units and thus cannot be a subclass of `power generating unit` . 

Seems, this inference goes as follows:
1. A power generating is something that has a generator as parts.
2. A power plant is something that has power generating units as parts and therefore also has generators as parts.w
3. As a power plant has generators as parts, it must be a power generating unit.

Seems we need here a different solution for `power generating unit`. I was not sure whether the equivalent class would work; that's why I've put a "maybe" in my last comment.

l-emele commented:
> Seems, this inference goes as follows:
> 
>     1. A power generating is something that has a generator as parts.
> 
>     2. A power plant is something that has power generating units as parts and therefore also has generators as parts.w
> 
>     3. As a power plant has generators as parts, it must be a power generating unit.

I just that this is because the object property `has part` is **transitive** why means that it "propagates", see: http://protegeproject.github.io/protege/views/object-property-characteristics/

For the parallel reasoning, the `solar thermal collector` parts make now that a `solar thermal power plant` and `solar thermal power unit` are now inferred subclasses of `heat generating unit`:
![grafik](https://user-images.githubusercontent.com/36884905/204356212-b62ab1ba-4b54-4296-9f59-099ac6d8b175.png)

Maybe we need to rethink the definitions (and equivalence axioms) of `power generating unit` and `heat generating unit` as being `energy transformation units` that have specific parts that differentiates them.

l-emele commented:
To say something positive about your large implementation efforts: The following classes `rotary heat exchanger`,  `plate heat exchanger`, `boiler`, `geothermal heat unit`, `geothermal heat plant`, `tube collector` and `flat plate collector` look fine and could be approved, if would not have the problems with the other classes and axioms.

Alex2448 commented:
> To say something positive about your large implementation efforts: The following classes `rotary heat exchanger`, `plate heat exchanger`, `boiler`, `geothermal heat unit`, `geothermal heat plant`, `tube collector` and `flat plate collector` look fine and could be approved, if would not have the problems with the other classes and axioms.

One idea is to move this problem in another issue and implement the rest, or maybe discuss it in the next dev meeting.

Alex2448 commented:
> > Seems, this inference goes as follows:
> > ```
> > 1. A power generating is something that has a generator as parts.
> > 
> > 2. A power plant is something that has power generating units as parts and therefore also has generators as parts.w
> > 
> > 3. As a power plant has generators as parts, it must be a power generating unit.
> > ```
> 
> I just that this is because the object property `has part` is **transitive** why means that it "propagates", see: http://protegeproject.github.io/protege/views/object-property-characteristics/
> 
> For the parallel reasoning, the `solar thermal collector` parts make now that a `solar thermal power plant` and `solar thermal power unit` are now inferred subclasses of `heat generating unit`: ![grafik](https://user-images.githubusercontent.com/36884905/204356212-b62ab1ba-4b54-4296-9f59-099ac6d8b175.png)
> 
> Maybe we need to rethink the definitions (and equivalence axioms) of `power generating unit` and `heat generating unit` as being `energy transformation units` that have specific parts that differentiates them.

@l-emele I tried to add some other restricitions in the equivalent to section of the `heat generating unti` but I cannot solve this I think.

@OpenEnergyPlatform/oeo-domain-expert-energy-modelling any idea how to solve this or what to do next? I can maybe for now implement `combined heat and power generating unit` as a subclass of both `heat generating unit` and `power generating unit` , we already have examples where we have 2 parent classes.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> 1. A power generating [unit] is something that has a generator as parts.
> 2. A power plant is something that has power generating units as parts and therefore also has generators as parts.
> 3. As a power plant has generators as parts, it must be a power generating unit.

Easy: 

1. A power generating unit is something that has a generator as parts and that does not have a grid component as parts.
2. A power plant is something that has power generating units as parts and has a grid component as parts, and therefore has a generator as parts.
3. As a power plants has a generator as parts and grid components as parts, it cannot be a power generating unit.

If they are supposed to be different, write down why they are different.

Alex2448 commented:
> > 1. A power generating [unit] is something that has a generator as parts.
> > 2. A power plant is something that has power generating units as parts and therefore also has generators as parts.
> > 3. As a power plant has generators as parts, it must be a power generating unit.
> 
> Easy:
> 
> 1. A power generating unit is something that has a generator as parts and that does not have a grid component as parts.
> 2. A power plant is something that has power generating units as parts and has a grid component as parts, and therefore has a generator as parts.
> 3. As a power plants has a generator as parts and grid components as parts, it cannot be a power generating unit.
> 
> If they are supposed to be different, write down why they are different.

The main problem we have is the `combined heat and power plant`. This is implemented in the dev branch as a subclass of `power plant`. But now, we have also `heat plant` and `combined heat and power plant` have parts of heat plant and power plant.  @l-emele proposed to make `heat plant` and `power plant` both equal to `energy transformation unit` but then we got problems (see above). One proposal is to make `combined heat and power plant` a subclass of both classes.

l-emele commented:
> But now, we have also `heat plant` and `combined heat and power plant` have parts of heat plant and power plant.

A `combined heat and power plant` a has `combined heat and power unit` as part (and via its parent) a grid connection, see also my [comment](https://github.com/OpenEnergyPlatform/ontology/pull/1362#discussion_r1039328937) in #1362.

Maybe the label `power generating unit` gives a direction how to redefine this class: _A power generating unit is an energy transformation unit that converts other forms of energy to electrical energy._
* `'power generating unit' EquivalentTo: 'energy transformation unit'  and 'has energy output' some 'electrical energy' and ('has energy input' some energy and 'has energy input' not some 'electrical energy'`
* `'power generating unit' SubclassOf: 'has part' some generator`

Similarly, we could redefine `heat generating unit` as: _A heat generating unit is an energy transformation unit is an energy transformation unit that converts other forms of energy to thermal energy._
* `'heat generating unit' EquivalentTo: 'energy transformation unit'  and 'has energy output' some 'thermal energy' and ('has energy input' some energy and 'has energy input' not some 'electrical energy'`
* `'heat generating unit' SubclassOf: 'has part' some (heater or 'heat exchanger)`

Then, to get `combined heat and power generating unit` as an inferred subclass of both `power generating unit` and `heat generating unit`, we have to add an energy input axiom. This seems fitting:
* `'combined heat and power generating unit' SubclassOf: 'has energy input' some 'chemical energy'`

l-emele commented:
I've put this PR on the agenda for the next OEO dev meeting: https://etherpad.wikimedia.org/p/oeo-dev-50

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
I don't get the "a power plant is an energy transformation unit" deal, seems too recursive to me.
> A _X_ is an energy transformation unit that consists of energy transformation units …

Wasn't the point of the "energy transformation units" that power (heat) plants can have multiple generating units (de: "Blöcke"), that some ESMs would describe separately?

If we need a super class to treat power and heat plants collectively, add an extra one (energy transformation plant?), instead of recycling something else.
Or is there a compelling reason for power plants to be the same "thing" as their components that I am missing?

-----

I don't see any problem with CHP plants being both power and heat plants, feels natural.  As a modeller, when I say "all power plants", I expect CHP plants to be included, unless I explicitly exclude them.  Furthermore, I think that relation should be inferred rather than explicit.

- Power plants have power generating units and electric grid components.
- Heat plants have heat generation units and heat grid components.
- Combined heat and power plants have combined heat and power generating units, electric grid components, and heat grid components.

- power generating unit
  - is a subclass of energy transformation unit
  - has energy output some electrical energy
  - has part some generator
  - participates in some energy transformation (N.B. could be more specific a "electricity generation process")
  - participates in some curtailment

- heat generating unit
  - is a subclass of energy transformation unit
  - has energy output some thermal energy
  - has part some (heat exchanger or heater)
  - participates in some heat generation process

So if we implement a combined power and heat generating unit that
- is a subclass of energy transformation unit
- has energy output some electrical energy
- has energy output some thermal energy
- has part some generator
- has part some (heat exchanger or heater)
- participates in some heat generation process (includes energy transformation)
- participates in some curtailment

that is both a "power generating unit" and a "heat generating unit", and subsequently something that has "combined heat and power generating units" and electrical and heat grid components would be both a power plant and a heat plant. as desired.

l-emele commented:
> > To say something positive about your large implementation efforts: The following classes `rotary heat exchanger`, `plate heat exchanger`, `boiler`, `geothermal heat unit`, `geothermal heat plant`, `tube collector` and `flat plate collector` look fine and could be approved, if would not have the problems with the other classes and axioms.
> 
> One idea is to move this problem in another issue and implement the rest, or maybe discuss it in the next dev meeting.

To move forward, I copy those undisputed classes into a new branch and create a new PR where I put you @Alex2448 as reviewer.

l-emele commented:
@Alex2448 : After merging #1432 into the dev branch, I merged the dev branch into this feature branch feature-1254-adding-more-heat-terms. This required some manual changes also in oeo-shared.

Now this PR here looks much cleaner.

chrwm commented:
I'll look into this PR on behalf of @Alex2448 by the end of the week.

l-emele commented:
We discussed this pull request in today's OEO dev meeting 50 and want to solve the problems by introducing functions. I created issue #1438 where we can the definitions of these functions.

l-emele commented:
We did not yet came to new definitions for `power generating unit`, `heat generating unit` and `combined heat and power generating unit`, but the definitions should be something like:
* _A power generating unit is an energy transformation unit that has the function to produce electrical energy._
* _A heat generating unit is an energy transformation unit that has the function to produce thermal energy._
* _A combined heat and power generating unit is an energy transformation unit that has the function to produce both thermal energy and electrical energy._

Further, we talked in OEO dev 50 about introducting a power-only-generating unit. But not to overload this PR, I'll create a separate issue for that.

stap-m commented:
> _A combined heat and power generating unit is an energy transformation unit that has the function to produce both thermal energy and electrical energy._

Since we have two different functions, I'd say: _"... that has both the function to produce thermal energy and the function to produce electrical energy."_

l-emele commented:
> > _A combined heat and power generating unit is an energy transformation unit that has the function to produce both thermal energy and electrical energy._
> 
> Since we have two different functions, I'd say: _"... that has both the function to produce thermal energy and the function to produce electrical energy."_

I like this improvement.

l-emele commented:
As discussed, I added the new function-based definitions to `power generating unit`, `heat generating unit` and `combined heat and power generating unit` (and merged the dev branch into this feature branch).

I'll have a close look now, whether all problems with this PR are solved now.

l-emele commented:
I think, we have solved now everything regarding this PR. So I will set it again to "Ready for review"!

chrwm commented on src/ontology/edits/oeo-shared.omn:
```diff
@@ -1642,6 +1650,9 @@ pull request: https://github.com/OpenEnergyPlatform/ontology/pull/810"
         <http://purl.obolibrary.org/obo/RO_0000087> some OEO_00040011
     
     
+Class: OEO_00000144
```
Should this be deleted?

l-emele replied:
No, it indicates that this class has to be moved to oeo-shared.

l-emele replied:
Okay, I resolved this. However, it implied to move some further classes to oeo-shared.

chrwm commented on src/ontology/edits/oeo-shared.omn:
```diff
@@ -2488,6 +2507,9 @@ pull request: https://github.com/OpenEnergyPlatform/ontology/pull/1390",
         OEO_00000014
     
     
+Class: OEO_00010386
```
Should this be deleted?

l-emele replied:
No, it indicates that this class has to be moved to oeo-shared.

chrwm reviewed (commented):
Thank you for pulling this together! 
I found two empty classes that might need deletion, and have one comment regarding the definition of parts for `heat generating unit`.

l-emele commented on src/ontology/edits/oeo-physical.omn:
```diff
@@ -9368,6 +9393,27 @@ pull request: https://github.com/OpenEnergyPlatform/ontology/pull/1359",
+    SubClassOf: 
+        OEO_00020102,
+        OEO_00010235 some OEO_00000207,
+        <http://purl.obolibrary.org/obo/BFO_0000051> some 
```
I see your point. However, I think we should keep the axiom `'heat transfer unit' (not ('has part' some heater))
 and ('has part' some 'heat exchanger')` as the definition explicitly mentions also the heater: _A heat transfer unit is a heat generating unit that contains a heat exchanger **and not a heater**._

l-emele replied:
The axiom `'solar heat unit' 'has part' some heater` is not necessary as there is already the axiom `'solar heat unit' 'has part' some 'solar thermal collector'` and `solar thermal collector` is a subclass of `heater`.

l-emele reviewed (approved):
I approve this PR now. However, as I was involved in finalising this PR it would be nice to get an additional approval.

areleu commented:
> I approve this PR now. However, as I was involved in finalising this PR it would be nice to get an additional approval.

I'll take a look tomorrow morning.

stap-m commented:
I approved this PR. 
We could think about a function for energy transfer and then take a look at the grid-related classes again (in another issue/PR😁).

l-emele commented:
Thanks for the review.
